### DISCUSSION TITLE
Good way to refresh VectorTileSource tiles by extent?

### DISCUSSION THREAD

**Discussion by Millcrest-Corey** [openlayers/openlayers#16220] [Q&A]
Does anyone have a good way to refresh specific tiles in a `VectorTileSource` based on extents? So instead of refreshing the whole source/layer; only the tiles that intersect with the specified extents are refreshed.

I've been using this messy workaround but it broke with v10.2.0 (specifically #16157) since it uses the `tileCache` that was removed.
```ts
private updateTimes: Record<string, number> = {}

public refreshTilesByExtents(extents: Extent[]) {
	if (this.source == null || extents.length <= 0) return

	const tileGrid = this.source.getTileGrid()
	if (tileGrid == null) return
	const tileCache = this.source["tileCache"]

	if (extents.some((extent) => getWidth(extent) > 500)) {
		this.source.refresh() 		// Refresh all tiles
		return
	}

	// Get unique z values from tile cache (so we don't have to loop over minZoom to maxZoom, ie: 0-23)
	const cacheKeys = Object.keys(tileCache["entries_"])
	const cachedZs = Array.from(
		new Set(cacheKeys.map((ck) => +ck.split("/")[0])),
	)

	let updated = false
	for (const extent of extents) {
		for (const z of cachedZs) {
			tileGrid.forEachTileCoord(extent, z, (tileCoord) => {
				const key = `${tileCoord[0]}/${tileCoord[1]}/${tileCoord[2]}`
				if (!tileCache.containsKey(key)) return

				// TODO: Is there a way we can refresh/load tiles without clearing them first? (so they don't flash black)
				this.updateTimes[key] = new Date().getTime()
				tileCache.remove(key)
				updated = true
			})
		}
	}
	if (updated) this.source.changed()
}
```

The `updateTimes` is then appended to the tileURL to force it to refetch:
```ts
this.source = new VectorTileSource<Feature>({
	format: new MVT({ idProperty: "id" }),
	tileUrlFunction: ([z, x, y]) => {
		const ut = this.updateTimes[`${z}/${x}/${y}`]
		return `${TI_TIPG_URL}/collections/${this.collection}/tiles/${tileMatrixSetId}/${z}/${x}/${y}${ut ? `?ut=${ut}` : ""}`
	},
	tileLoadFunction: this.tileLoadFunction,
	projection,
	maxZoom: 23,
	tileGrid: new TileGrid(tileGrid),
	transition: 0,
	cacheSize: 0,
})
```

**Comment by ahocevar** [comment]
Why do you want to do that? For vector tiles, the tile cache is usually very small (2 zoom levels of tiles fitting the current viewport) to avoid canvas memory issues. Why don't you just call `refresh()` on the source?

**Reply by Millcrest-Corey** [reply]
When I call `refresh()` it hits the endpoint around 12-20 times because it's refreshing multiple tiles, even though only a couple have any changes. With my current technique it only refetches 1 or 2 which is a lot better for performance as random tiles can refresh quite often.

It has quite a lot of zoom levels of tiles since simplification is dynamic based on the zoom/resolution and passed in the tileURL as well (not shown in the above example).

**Reply by ahocevar** [reply]
This sounds like your server is not sending appropriate cache headers. If it would, the browser cache would be used, instead of sending new requests.
